### PR TITLE
Add support for AVG() on INTERVAL types

### PR DESCRIPTION
**Pull request by matriv**
Closes: #13677

### CODE REVIEW

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/average/IntervalAverageAggregation.java [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+
+    @Override
+    public IntervalAverageState iterate(RamAccounting ramAccounting,
+                                        MemoryManager memoryManager,
+                                        IntervalAverageState state,
+                                        Input<?>... args) {
+        if (state != null) {
+            Period value = (Period) args[0].value();
+            if (value != null) {
+                state.addPeriod(value); // Mutates state.
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/average/IntervalAverageAggregation.java:215]: Can we add a test that avg of 2 nulls is null? It will also cover ` if (count > 0) { ... else return null`

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/average/IntervalAverageAggregation.java [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+
+    public static class IntervalAverageState implements Comparable<IntervalAverageState> {
+
+        private Period sum = new Period();
+        private long count = 0;
+
+        public Period value() {
+            if (count > 0) {
+                return new Period(sum.toStandardDuration().getMillis() / count);
+            } else {
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/average/IntervalAverageAggregation.java:80]: Just a question: is it aligned with PG?
There can be another ways:
1. this approach but similar to `age`, normalize output like 
`new Period(sum.toStandardDuration().getMillis() / count).normalizedStandard(PeriodType.yearMonthDayTime())`
`PeriodType.yearMonthDayTime()` is aligned with PG default representation but not sure it's relevant here for avg.

2. computing average per unit. - `new Period().days(cureDays/count).hours(curHours./count)` and so on except months and years which shouldn't be  among units anyway as it's checked in `toStandardDuration()` internally via `checkYearsAndMonths` - but I think this is overkill.

To sum up, this approach looks good but I think we need to add the following things to the docs:

1. `toStandardDuration() `internally throws an exception if Period has months/year units - maybe we need to add an example and mention that in docs that it will throw an error. Maybe we could also check same example in PG 

2. add a working example and maybe explicitly mention that avg is NOT avg per units (ie. NOT approach 2)

File: server/src/test/java/io/crate/execution/engine/aggregation/impl/AverageAggregationTest.java [hunk 3]
```diff
@@ -65,6 +66,18 @@ private Object executeAggregation(DataType<?> argumentType, Object[][] data) thr
         );
     }
 
+    private Object executeIntervalAggregation(Object[][] data) throws Exception {
+        return executeAggregation(
```
[the remark this answers is not in the document]
Comment (matriv) [reply server/src/test/java/io/crate/execution/engine/aggregation/impl/AverageAggregationTest.java:70]: It's fine, it will be more clear, will rename both methods.

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/average/IntervalAverageAggregation.java (hunk 2 above)
Comment (matriv) [reply server/src/main/java/io/crate/execution/engine/aggregation/impl/average/IntervalAverageAggregation.java:80]: Thanks a lot for diving into this @BaurzhanSakhariev !
- PostgreSQL uses 1st approach (not an avg per unit but a total average in millis)
- I will use our custom `toStandardDuration()` which works for months and years with 30 and 365 days respectively.

Comment (matriv) [reply server/src/main/java/io/crate/execution/engine/aggregation/impl/average/IntervalAverageAggregation.java]: postgres doesn't normalize to months/years but only up to days, for avg();
I replicated the integ test in psql:
```
matriv=> SELECT avg('2023-05-04 22:33:11.123'::timestamp - ts) FROM tbl;
          avg           
------------------------
 125 days 25:42:24.4535
(1 row)

```

**Comment by matriv** [comment]
@BaurzhanSakhariev Please check again.
Apologies, there were many changes because of the fixes so, I squashed things and force pushed.

File: server/src/main/java/io/crate/types/IntervalType.java [hunk 4]
```diff
@@ -181,12 +182,29 @@ public long valueBytes(Period value) {
...
+        millis = millis.add(BigInteger.valueOf(p.getSeconds() * (long) DateTimeConstants.MILLIS_PER_SECOND));
+        millis = millis.add(BigInteger.valueOf(p.getMinutes() * (long) DateTimeConstants.MILLIS_PER_MINUTE));
+        millis = millis.add(BigInteger.valueOf(p.getHours() * (long) DateTimeConstants.MILLIS_PER_HOUR));
+        millis = millis.add(BigInteger.valueOf(p.getDays() * (long) DateTimeConstants.MILLIS_PER_DAY));
+        millis = millis.add(BigInteger.valueOf(p.getWeeks() * (long) DateTimeConstants.MILLIS_PER_WEEK));
+        millis = millis.add(BigInteger.valueOf(p.getMonths() * 30 * (long) DateTimeConstants.MILLIS_PER_DAY));
+        millis = millis.add(BigInteger.valueOf(p.getYears() * 12 * 30 * (long) DateTimeConstants.MILLIS_PER_DAY));
+        return millis;
+    }
+
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/types/IntervalType.java:201]: Minor:
 
 I think we can slightly reduce number of allocations per addition:
 
 `Period.get***` methods, even `getMillis` return `int`. 
 Mills in a week is ~10^8 (makes it 10^17 when multiplied by int), so up to weeks we are fine with long:
 to overflow we need 100 more 10^17 magnitude summands but all units up to weeks are much smaller because DateTimeConstants values are small.
 
We can use our `OverflowAwareMutableLong` which has logic "use primitive as much as possible, flush when needed to BigDecimal, then reset and again start with primitives".  Since it's BigDecimal, we would need OverflowAwareMutableLong.value().toBigInteger() at the end but it will save us those 5 allocations

**Review by BaurzhanSakhariev — APPROVED** [review]
👍 Thanks!

Left a minor comment

File: server/src/main/java/io/crate/types/IntervalType.java (hunk 4 above)
Comment (matriv) [reply server/src/main/java/io/crate/types/IntervalType.java]: Thx @mfussenegger! I will look more into this, and see if I can follow up with improvement.
